### ISSUE TITLE
Prisma binding class should not use global variables

### ISSUE THREAD

**Issue by kbrandwijk** [prisma/prisma-binding#81]
`typeDefsCache` and `remoteSchema` are currently variables outside the Prisma class (making them global). This is done as a form of 'caching', so generating the remote schema does not happen on every request.

However, at the same time, this makes it impossible to use more than one Prisma binding, because the global variables will be shared across those.

Bindings should be completely isolated from one another.

**Comment by schickling** [comment]
Great point! This should definitely be improved. The caching mechanism is intended however it should cache based on a schema level (so instead of a single cache variable there should be a cache map).

**Comment by a-type** [comment]
Newcomer to the project; Git history shows `graphql-schema-cache` previously used. Was there a reason for dropping it?

**Comment by schickling** [comment]
Yes this was based on a different „architecture“ and is now no longer needed. It refers to another kind of cache than this issue describes.

**Comment by a-type** [comment]
Great, I think that's enough info for me to get started on a PR.

**Comment by pollard93** [comment]
Another newcomer! I have another use case: i'm building a multi tenant system of which I am deploying multiple databases with identical schemas. E.g service-1@dev and service-2@dev.

I want to then switch the endpoint when creating a new Prisma instance on each request. E.g endpoint/service-1/dev and endpoint/service-2/dev.

However due to `getCachedRemoteSchema` using the `typeDefs` as the key, it's impossible to switch endpoints on the fly if the schemas are identical.

I've got it working by using the endpoint as the key. Is this a suitable solution for my use case?

**Comment by Errorname** [comment]
Hi there!
I have the same issue as @pollard83, I can't switch between multiple endpoints having the same schema.

But I've found a solution which doesn't require me to fork this repo until a better solution is found.

I currently am doing a workaround by importing the typeDefs from the file myself and then appending a comment with the service and stage names. This makes the cache think the typedefs are different.

**Comment by pollard93** [comment]
@Errorname Nice workaround I'm now doing the same, thanks.

**Comment by maticzav** [comment]
Hey 👋, I believe this issue has been resolved in most recent versions of `prisma-binding`. Taking @schickling's comment as a basis, I believe new implementation constructs a remote schema based on provided `typeDefs` instead of blindly saving a global default.

More about this can be found in the `cache.ts` file of `prisma-binding`; I'll include the link below.

https://github.com/prisma/prisma-binding/blob/master/src/cache.ts

**Comment by Errorname** [comment]
Hey @maticzav !
Well, @pollard93 andI still have an issue with the current implementation. Here is our issue:

We both have a multi tenant system in which we created multiple services having the exact same schemas (and therefore same provided typeDefs). Then, using prisma-binding, we try to register some (at least two) of our identical services. But, **because they have the same typeDefs**, the cache only sends requests to the first registered service!

In order to make a multi-tenant application with prisma-binding, we have to bypass the pass using the following trick:
For each of our services, we manually import the typeDefs file ourselves, and append a comment with the service and stage name. The cache is now fooled thinking they are different typeDefs and allow us to query independently our services.

I hope I explained why the cache is *still* a issue, at least for our use-case.
